genie-agent/src/main/java/com/netflix/genie/agent/execution/services/LaunchJobServiceImpl.java

```diff
@@ -0,0 +1,153 @@
+
+        configureEnvironment(processBuilder);
+        configureCommandArguments(processBuilder, jobSpec);
+        configureIO(processBuilder, jobSpec);
+
+        try {
+            return processBuilder.start();
+        } catch (final IOException | SecurityException e) {
+            throw new JobLaunchException("Failed to launch job: ", e);
+        }
+    }
+
+    private void configureEnvironment(
+        final ProcessBuilder processBuilder
+    ) throws JobLaunchException {
+
+        final File jobDirectory = executionContext.getJobDirectory();
+
+        // Validate job running directory
+        if (jobDirectory == null) {
+            throw new JobLaunchException("Job directory is not set in execution context");
+        } else if (!jobDirectory.exists()) {
+            throw new JobLaunchException("Job directory does not exist: " + jobDirectory);
+        } else if (!jobDirectory.isDirectory()) {
+            throw new JobLaunchException("Job directory is not a directory: " + jobDirectory);
+        } else if (!jobDirectory.canWrite()) {
+            throw new JobLaunchException("Job directory is not writeable: " + jobDirectory);
+        }
+
+        // Configure job running directory
+        processBuilder.directory(jobDirectory);
+
+        final Map<String, String> currentEnvironmentVariables = processBuilder.environment();
+        final Map<String, String> jobEnvironmentVariables = executionContext.getJobEnvironment();
+
+        // Validate additional environment variables
+        if (jobEnvironmentVariables == null) {
+            throw new JobLaunchException("Job environment variables map is not set in execution context");
+        }
+
+        // Configure job environment variables
+        jobEnvironmentVariables.forEach((key, value) -> {
+            final String replacedValue = currentEnvironmentVariables.put(key, value);
+            if (StringUtils.isBlank(replacedValue)) {
+                log.debug(
+                    "Added job environment variable: {}={}",
+                    key,
+                    value
+                );
+            } else if (!replacedValue.equals(value)) {
+                log.debug(
+                    "Set job environment variable: {}={} (previous value: {})",
+                    key,
+                    value,
+                    replacedValue
+                );
+            }
+        });
+    }
+
+    private void configureCommandArguments(
+        final ProcessBuilder processBuilder,
+        final JobSpecification jobSpec
+    ) throws JobLaunchException {
+        final List<String> commandLineArguments = jobSpec.getCommandArgs();
+
+        // Validate arguments
+        if (commandLineArguments == null) {
+            throw new JobLaunchException("Job command-line arguments not set in execution context");
+        } else if (commandLineArguments.isEmpty()) {
+            throw new JobLaunchException("Job command-line arguments are empty");
```

Comment:
I think this is fine and shouldn't throw an exception. The reason I made commandArgs as a non-required field in Genie 3.3 and soon 4.0 was cause I could imagine a command that is just like `/spark-shell` and no command line arguments are required to dump you into a default shell. In the past we had to get around this but putting a space in as the only command line argument which was just a waste.

Reply:
It depends, does args include the executable?
I assumed it did, since it's not sent back in the job spec in any other place.
(And I think it's good to have the entire command-line in one parameter in the spec object, as opposed to have the binary somewhere else hard to find)

Reply:
oh this is a good point. I totally missed doing this on the server side of the job spec. Let me fix that bug. I'll pass it back in this command args list then this check is accurate. The arguments should NEVER be null cause the DTO doesn't return a null though.

Reply:
#704 addresses this